clients/src/main/java/org/apache/kafka/common/errors/AuthenticationFailedException.java

```diff
@@ -16,6 +16,19 @@
  */
 package org.apache.kafka.common.errors;
 
+/**
+ * This exception indicates that SSL handshake or SASL authentication has
+ * failed.
+ * <p>
+ * SSL handshake failures in clients may indicate client authentication
+ * failure due to untrusted certificates if server is configured to request
+ * client certificates. Handshake failures could also indicate misconfigured
+ * security including protocol/cipher suite mismatch, server certificate
+ * authentication failure or server host name verification failure.
+ * </p><p>
+ * SASL authentication failures typically indicate invalid credentials.
+ * <p>
+ */
 public class AuthenticationFailedException extends AuthenticationException {
```

Comment:
I wonder if this should be called `InvalidCredentialsException`.

Reply:
At the moment, we have a single error code and exception for authentication failure with an error message giving more details. For SSL, this includes any handshake exception (e.g. server host name verification failure). At the moment, we are reporting only invalid credentials for SASL, but we may add others too. Since we convert SSLException and SaslException from the providers, it is simpler to maintain a more generic `AuthenticationFailed` exception rather than separate exceptions for different failures.

Reply:
The issue I have is that `AuthenticationFailedException` is too similar to `AuthenticationException`.

Reply:
What about `SaslAuthenticationException` and `SslAuthenticationException`?

Reply:
That sounds good to me.

Reply:
I will add without the `Failed` to be consistent with `TopicAuthorizationException` etc. So `SaslAuthenticationException` and `SslAuthenticationException`.

Reply:
I read it a bit too quickly on my phone and I hadn't seen the "Failed" part. What you did in the end is what I had in mind.

Reply:
I removed the SSL exception part from the doc. Will add `SslAuthenticationException` and the associated docs in #3918